### PR TITLE
Refactor exporter - step 6

### PR DESCRIPTION
**Pull request by reyang**
This is a follow up PR of #1078.

## Changes

* Implemented `BatchExportActivityProcessor`.
* Updated the tutorial to reflect the changes.
* Dispose/Shutdown issue will be addressed in a separate PR. Ref: #1024.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 1]
```diff
@@ -127,6 +133,82 @@ public override void OnEnd(Activity activity)
...
+
+                    if (this.shutdownTrigger.WaitOne(0))
+                    {
+                        return false;
+                    }
+                }
+            }
+            finally
+            {
+                sw.Stop();
```
Comment (pjanotti) [inline src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:208]: You don't need to call `Stop()` for a stopwatch that you are not going to use anymore. You can get rid of the try-finally.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 2]
```diff
@@ -165,5 +291,29 @@ protected override void Dispose(bool disposing)
...
+                // only wait when the queue doesn't have enough items, otherwise keep busy and send data continuously
+                if (this.queue.Count < this.maxExportBatchSize)
+                {
+                    WaitHandle.WaitAny(triggers, this.scheduledDelayMillis);
+                }
+
+                this.exporter.Export(new Batch<Activity>(this.queue, this.maxExportBatchSize));
+
+                this.dataExportedNotification.Set();
+                this.dataExportedNotification.Reset();
```
Comment (pjanotti) [inline src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs]: Is this guaranteed to wake up the one waiting?

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 3]
```diff
@@ -165,5 +291,32 @@ protected override void Dispose(bool disposing)
...
+                {
+                    WaitHandle.WaitAny(triggers, this.scheduledDelayMillis);
+                }
+
+                if (this.queue.Count > 0)
+                {
+                    this.exporter.Export(new Batch<Activity>(this.queue, this.maxExportBatchSize));
+
+                    this.dataExportedNotification.Set();
+                    this.dataExportedNotification.Reset();
```
Comment (pjanotti) [inline src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:312]: Is this guaranteed to wake-up one waiting for the signal?

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 4]
```diff
@@ -127,6 +137,75 @@ public override void OnEnd(Activity activity)
...
+        /// </returns>
+        public bool ForceFlush(int timeoutMillis = Timeout.Infinite)
+        {
+            if (timeoutMillis < 0 && timeoutMillis != Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMillis));
+            }
+
+            var tail = this.queue.RemovedCount;
+            var head = this.queue.AddedCount;
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:161]: Yes it is.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 5]
```diff
@@ -29,9 +29,13 @@ public class BatchExportActivityProcessor : ActivityProcessor
     {
         private readonly ActivityExporterSync exporter;
         private readonly CircularBuffer<Activity> queue;
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:31]: Good point, renamed.

[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs]: After the change it seems we're still quite messy with name, as the spec is asking for `maxQueueSize`, so we have something weird like `new CircularBuffer<Activity>(maxQueueSize)`.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 6]
```diff
@@ -29,9 +29,13 @@ public class BatchExportActivityProcessor : ActivityProcessor
...
-        private readonly TimeSpan scheduledDelay;
-        private readonly TimeSpan exporterTimeout;
+        private readonly int scheduledDelayMillis;
+        private readonly int exporterTimeoutMillis;
         private readonly int maxExportBatchSize;
+        private readonly Thread exporterThread;
+        private readonly ManualResetEvent exportTrigger = new ManualResetEvent(false);
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:36]: I don't see the slim version giving benefit other than burning CPU cycles.
Most of these operations will wait for I/O operation which takes longer time so the spin will upgrade to a real event.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 7]
```diff
@@ -127,6 +133,75 @@ public override void OnEnd(Activity activity)
             Interlocked.Increment(ref this.droppedCount);
         }
 
+        /// <summary>
+        /// Flushes the <see cref="Activity"/> currently in the queue, blocks
+        /// the current thread until flush completed, shutdown signaled or
+        /// timed out. Using a 32bit signed integer to specify the time
+        /// interval in milliseconds.
+        /// </summary>
+        /// <param name="timeoutMillis">
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:145]: 1. It aligns better with the spec.
2. It allows default parameter to be specified in the signature directly, which results in simpler code.
3. Debugging a crash dump would be simpler.

None of the above gives any significant benefit though.

File: docs/trace/building-your-own-exporter/Program.cs [hunk 8]
```diff
@@ -35,5 +37,21 @@ public static void Main()
...
+
+        using (var activity = MyActivitySource.StartActivity("SayHello2"))
+        {
+        }
+
+        processor.ForceFlush();
+        processor.ForceFlush();
+        processor.ForceFlush();
+        processor.ForceFlush();
+        processor.ForceFlush();
```
Comment (CodeBlanch) [inline docs/trace/building-your-own-exporter/Program.cs]: There is a 🚽 joke to be made here, but I don't want to be a drain.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 9]
```diff
@@ -29,9 +29,13 @@ public class BatchExportActivityProcessor : ActivityProcessor
     {
         private readonly ActivityExporterSync exporter;
         private readonly CircularBuffer<Activity> queue;
-        private readonly TimeSpan scheduledDelay;
-        private readonly TimeSpan exporterTimeout;
+        private readonly int scheduledDelayMillis;
+        private readonly int exporterTimeoutMillis;
         private readonly int maxExportBatchSize;
+        private readonly Thread exporterThread;
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:35]: I think having an explicit thread gives a lot of benefit:
1. The exporter thread runaway time would give idea on how much CPU cycles were used by exporting job.
2. While debugging a live process or a dump, it is easy to see whether the exporting job got stuck or not.
3. It avoids the accidental context leakage of `SuppressInstrumentation`.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs [hunk 10]
```diff
@@ -165,5 +289,32 @@ protected override void Dispose(bool disposing)
...
+                {
+                    WaitHandle.WaitAny(triggers, this.scheduledDelayMillis);
+                }
+
+                if (this.circularBuffer.Count > 0)
+                {
+                    this.exporter.Export(new Batch<Activity>(this.circularBuffer, this.maxExportBatchSize));
+
+                    this.dataExportedNotification.Set();
+                    this.dataExportedNotification.Reset();
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:310]: I think we also need a `this.exportTrigger.Reset();` here or it's going to be stuck in the signaled state. Alternatively, AutoResetEvent should work for `exportTrigger`. I'm pretty sure it stays signaled until at least one thing is triggered, which in our case would be an export.

Comment (reyang) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessor.cs:310]: Nice catch, that line was missed while I was cleaning up the code.

**Pull request merged**